lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -0,0 +1,184 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./byte-efficiency-audit');
+const URL = require('../../lib/url-shim');
+
+const IGNORE_THRESHOLD_IN_BYTES = 2048;
+
+class UnusedJavaScript extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'unused-javascript',
+      description: 'Unused JavaScript',
+      informative: true,
+      helpText: 'Remove unused JavaScript to reduce bytes consumed by network activity.',
+      requiredArtifacts: ['JsUsage', 'networkRecords']
+    };
+  }
+
+  /**
+   * @param {{used: number, unused: number}} objA
+   * @param {{used: number, unused: number}} objB
+   * @return {{used: number, unused: number}}
+   */
+  static _add(objA, objB) {
+    return {
+      used: objA.used + objB.used,
+      unused: objA.unused + objB.unused,
+    };
+  }
+
+  /**
+   * Determines the length of unused and used portions in a set of ranges.
+   * Because the protocol reports usage for each function and ranges within each function, nested
+   * regions are represented numerous times. This function makes sure to only count each string
+   * section once. See unit and smoke tests for examples.
+   *
+   * @param {!Array<{isUsed: boolean, startOffset: number, endOffset: number
+   *    used: number, unused: number}>} ranges
+   * @return {{used: number, unused: number}}
+   */
+  static findUsedAndUnused(ranges) {
+    if (ranges.length === 0) {
+      return {used: 0, unused: 0};
+    } else if (ranges.length === 1) {
+      return ranges[0];
+    }
+
+    const firstRange = ranges[0];
+    const secondRange = ranges[1];
+    if (secondRange.startOffset >= firstRange.endOffset) {
+      // Ranges aren't nested, we can simply continue on to the rest.
+      return this._add(
+        this._add(firstRange, {used: secondRange.startOffset - firstRange.endOffset, unused: 0}),
+        this.findUsedAndUnused(ranges.slice(1))
+      );
+    } else {
+      let endIndex = 1;
+      while (endIndex < ranges.length && ranges[endIndex].startOffset < firstRange.endOffset) {
+        endIndex++;
+      }
+
+      const spanningRanges = ranges.slice(1, endIndex);
+      const restOfRanges = ranges.slice(endIndex);
+      if (firstRange.isUsed) {
+        // Nested ranges may potentially be unused which we need to account for.
+        const earliestStart = secondRange.startOffset;
+        const latestEnd = spanningRanges.reduce((max, next) => {
+          return max.endOffset >= next.endOffset ? max : next;
+        }).endOffset;
+        const nextStart = (restOfRanges[0] && restOfRanges[0].startOffset) || firstRange.endOffset;
```

Comment:
are ranges always nested? That is, if `secondRange.startOffset < firstRange.endOffset`, does that automatically imply that `secondRange.endOffset <= firstRange.endOffset`?